### ISSUE TITLE
heap overflow in check_literal()

### ISSUE THREAD

**Issue by hannob** [jqlang/jq#1234]
When passing certain files to jq it will cause a one byte write heap overflow. A simple example is just a file containing 256 times the same character.
E.g. create a file with:
for i in $(seq 1 256); do echo -n a; done > jq1
Then
jq . jq1

This bug was found with the tool american fuzzy lop.

jq doesn't crash, one needs to compile it with address sanitizer to see the bug. Here's the debugging output from address sanitizer:

```
==7113==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x61100001c700 at pc 0x0000005aa15f bp 0x7ffe0d461a70 sp 0x7ffe0d461a68
WRITE of size 1 at 0x61100001c700 thread T0
    #0 0x5aa15e in check_literal /mnt/ram/jq-1.5/jv_parse.c:488:30
    #1 0x5a27c6 in jv_parser_next /mnt/ram/jq-1.5/jv_parse.c:782:16
    #2 0x62dfc4 in jq_util_input_next_input /mnt/ram/jq-1.5/util.c:444:15
    #3 0x4fd1d7 in main /mnt/ram/jq-1.5/main.c:528:34
    #4 0x7f01ab56f78f in __libc_start_main /var/tmp/portage/sys-libs/glibc-2.23-r2/work/glibc-2.23/csu/../csu/libc-start.c:289
    #5 0x41acb8 in _start (/mnt/ram/jq-1.5/jq+0x41acb8)

0x61100001c700 is located 0 bytes to the right of 256-byte region [0x61100001c600,0x61100001c700)
allocated by thread T0 here:
    #0 0x4c2f58 in realloc (/mnt/ram/jq-1.5/jq+0x4c2f58)
    #1 0x6152ec in jv_mem_realloc /mnt/ram/jq-1.5/jv_alloc.c:162:7
    #2 0x5ad976 in tokenadd /mnt/ram/jq-1.5/jv_parse.c:388:19
    #3 0x5a7bd5 in scan /mnt/ram/jq-1.5/jv_parse.c:626:7
    #4 0x5a096d in jv_parser_next /mnt/ram/jq-1.5/jv_parse.c:743:11
    #5 0x62dfc4 in jq_util_input_next_input /mnt/ram/jq-1.5/util.c:444:15
    #6 0x4fd1d7 in main /mnt/ram/jq-1.5/main.c:528:34
    #7 0x7f01ab56f78f in __libc_start_main /var/tmp/portage/sys-libs/glibc-2.23-r2/work/glibc-2.23/csu/../csu/libc-start.c:289

SUMMARY: AddressSanitizer: heap-buffer-overflow /mnt/ram/jq-1.5/jv_parse.c:488:30 in check_literal
Shadow bytes around the buggy address:
  0x0c227fffb890: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c227fffb8a0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c227fffb8b0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c227fffb8c0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c227fffb8d0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0c227fffb8e0:[fa]fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
  0x0c227fffb8f0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c227fffb900: 00 00 fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c227fffb910: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c227fffb920: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c227fffb930: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==7113==ABORTING
```

**Comment by pkoppstein** [comment]
1.  Using the current ("development") version of jq:

    $ jq . jq1
    parse error: Invalid numeric literal at EOF at line 1, column 256

So I think the problem has already been solved.

2.  I think the bug is irrelevant anyway.  Since the input is not valid JSON, you'd normally read it as text, i.e. with the -R option.  Even jq 1.3 gets that right:

    $ jq1.3 -R . jq1
    "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"

3. Thanks for using american fuzzy lop!

**Comment by dequis** [comment]
Yeah, duplicate of #995 (8eb1367ca44e772963e704a700ef72ae2e12babd, CVE-2015-8863)

**Comment by hannob** [comment]
Ok, seems this got fixed in the repo. But if this already got a CVE last year - shouldn't there be a new release with the fix?

**Comment by itchyny** [comment]
Looks like the fix was released in [1.6](https://github.com/jqlang/jq/releases/tag/jq-1.6) in Nov 2018, so closing this issue.
